### PR TITLE
Addition of hourly temporal resolution_Continuous-time format

### PR DESCRIPTION
**Pull request by erikfilias**
While taking the fact that models contain data in temporal resolution, it so necessary to adopt a common datetime format to validate the data and synchronize efforts.

According to the feedbacks, we have the following:

- [x] Standard datetime format     : **UTC**  
> For example: `2020-01-01, 2020-01-02, ...` or `2020-01-01T13:00, ...`
> Inclusion of attributes is possible such as duration (More details [here](https://github.com/openENTRANCE/nomenclature/blob/master/nomenclature/definitions/subannual/months.yaml))
- [x] Using time zone                    : **no relevance**
- [x] Using winter/summer time   : 
> To distinguish between different granularity levels of representative timeslices, It was proposed the following: `<Granularity>|<Name of timeslice>`. For example: `2 Season-2 Times|Summer-Day`.
- [x] Averaging values over the time span   : 
> A value is always the average for flow variables (i.e. It is the average between the subannual time and the subsequent one, where average is contingent on the lowest level of granularity - if you use 2020-01-01T13:00, it is hourly average, if you use 2020-01-01, it is daily average...). 
> Reference comment -> [here](https://github.com/openENTRANCE/nomenclature/issues/46#issuecomment-641721712)
- [x] Accumulating values over the time span: 
> A value at the start of the period for stock over the time-period until the start of the next timeslice. 
> For example, Capacity or Reservoir Level at 2020-01-01T13:00 is the value at 1pm, if subannual is given as 2020-01-01, it is understood as midnight that day, if it's January, it is understood as midnight on the first day of the month. 
> Reference comment -> [here](https://github.com/openENTRANCE/nomenclature/issues/46#issuecomment-641721712)

##
In order to deal with this issue of a standard format, we'll make it in 2 PR:
This one related to continuous-time format, and new one related to representative timeslices.
##
Then, this PR consider an update of readme related to subannual column which we use to put the datetime format, and script to get a list of dates and hours.

### CODE REVIEW

**Comment by danielhuppmann** [comment]
Thanks @erikfilias for starting the discussion how to include continuous-time data in the common format and nomenclature.

A few issues that need clarification, in my opinion:
 - do we really want that each modelling team can use any time zone? Wouldn't it be easier for data exchange to use UTC throughout, to make data more easily comparable?
 - how are modelling teams supposed to work with summer-time?

Also, how about models that do not have hourly resolution but e.g. 12-hourly resolution:

 - how will a user know what the model/scenario time resolution is?
 - is the value for `2020-01-01T12:00+00:00` related to
   - the value in that hour
   - the average over the time span 12pm until 12am
   - the average over the time span 6am until 6pm
 -  again summer time: should this model report its data in summer time as `2020-01-01T13:00+00:00` during summer-time (to have consistent timesteps over the year), or should the averaging be adjusted on the time-shift days?

I know that ISO 8601 **can** represent any of these issues - what we need is more guidance for the modellers on how they **should** use it in the openENTRANCE context.

@tburandt @sandrinecharousset @arght would appreciate for your point of view

**Comment by sandrinecharousset** [comment]
In plan4eu we use UTC timestamps. I suppose converting UTC to the format proposed here should not be a problem. I had never thought of the timezones problem.... UTC seems enough to me as you can easily also get the timezone from the region and then convert into the proposed format. We just have to choose one.
regarding time granularity, here is what is done in plan4eu:
* the model can cope with diverse time granularity, either for different variables and also for the same variable. In simple datasets (as our public one), some of the variables have hourly granularity, others have weeky, monthly or yearly.
* the model can cope with variable length of timesteps. eg you can have for instance a variable where you will give one value for january, then one value per day for february.....
* the convention used is that one:
   - a value Va is given for time step ta, then another one Vb for timestep tb (t1 and t2 expressed as UTC)
   - this means that from ta until tb excluded, the value of the variable to be used is Va.
   - depending if Va is expressed as a power (eg MW) or as an energy (MWh), the data reading tool manages differently:
             + case of power: Va is duplicated on all instants between ta and tb excluded
             + case of energy: Va is used on ta, then all instants before tb have value 0 (this is only a choice in our inputs-formatting tool, we could also have chosen to spread out the value Va (eg if there are 3 hours between ta and tb and the value is 3MWH, put 1MWh on each hour instead of putting the 3MWh on ta and 0 on the 2 following timesteps)

* we do not deal with winter/summer time (this has been taken into account by the models generating the time profiles electricity demands as it has an impact but this model is out of scope)

**Comment by arght** [comment]
Hi all,
I coincide with the comments from Sandrine.
In openTEPES we deal with hourly data and we solve different time intervals just taking the variables at the end of those time intervals. For example, if you want to work with bi-hours (2-h timestep) the model activates only the variables/equations for the even hours. This is extended for any interval that is a multiple of meaninful timesteps (3-h, 4-h and even 6-h timesteps). 
In the same way we activate variables that have only meaning on a daily/weekñy/montly basis. We activate the variable only the last hour of a day/week/month.
openTEPES is a long-term planning model and therefore summer/winter time are not concerned. 
We don't deal with days with 23 or 25 hours like those where the time changes.
Best regards
Andrés

**Comment by arght** [comment]
No problem on our side on using ISO or UTC time definition
Andrés

**Comment by sandrinecharousset** [comment]
@arght we seem to have very close conventions for dealing with different duration timesteps except that in our case we use the first instant of a period. eg the value of a variable for a week (if weekly data) is usually defined on the monday (as we use UTC we do not care of the hour, just write eg 02/07/2043 ), and if the next changing value is only 2 weeks later then the following line in the dataset is 16/07/2043 and the model takes the value of 02/07/2043 for the week starting 09/07/2043

**Comment by arght** [comment]
Sandrine,
We formally define all the variables at the end of the corresponding period (hour, day, week, etc.). This is relevant in the inventory constraints (e.g., water balance) for defininng the running index.
inventory at the end of the previous period + inflows in this period - production in this period - spillage in this period = inventory at the end of the current period
Andrés

**Comment by sandrinecharousset** [comment]
@arght ok that makes sense. Our convention is to define variables at the time instant when the value changes. But we can also define them hourly (but then the volumetry of data is very big).

**Comment by arght** [comment]
Sandrine,
Regarding for example inflows, where we manage weekly constraints, we can define hourly inflows but also just a single inflow associated to a certain hour with the cumulative inflows for the whole week, which substanttially reduces the amount of data.
Andrés

**Comment by erikfilias** [comment]
Many thanks for the comments. I didn't say it but my purpose was about getting attention from modelling teams to this issue and get agreements to choose a datetime format.
 I'll be checking the comments and summarize the agreements.
The checks means that we talked about it.
Until now, we have the following points:

- [ ] Standard datetime format     : UTC
- [ ] Using time zone                    : no relevant
- [ ] Using winter/summer time   : no one use
- [ ] Using 12-hourly resolution   : What about use am & pm format?

I'll put winter/summer time issues if some modeling teams are interested.

**Comment by arght** [comment]
Erik,
I think that the 12-hourly resolution (two time periods per day, e.g., meaning on-peak/off-peak or day/night, morning/afternoon, midday/midnight) can be solved by the mechanisms that Sandrine and myself have proposed. Basically, picking up two individual hours as representative of those periods. Alternatively, if this is not enough specific names can be added as those mentioned before.
Andrés

**Comment by erikfilias** [comment]
> Erik,
> I think that the 12-hourly resolution (two time periods per day, e.g., meaning on-peak/off-peak or day/night, morning/afternoon, midday/midnight) can be solved by the mechanisms that Sandrine and myself have proposed. Basically, picking up two individual hours as representative of those periods. Alternatively, if this is not enough specific names can be added as those mentioned before.
> Andrés

Thank you, @arght. I got it and agree. This issue was not clear with `Using 12-hourly resolution : What about use am & pm format?`. I replaced it with  `Averaging over the time span`.
Until now, we have the following points:

- [x] Standard datetime format     : UTC
- [x] Using time zone                    : no relevant
- [x] Using winter/summer time   : no one use
- [x] Averaging over the time span   : Picking up an individual datetime as representative of time span

If there any detail that is not considered, don't hesitate to comment, please.

**Comment by danielhuppmann** [comment]
@erikfilias unfortunately, about summer-time, not all modelling teams are following this discussion on GitHub - did you check with all teams that they are not using it?

**Comment by danielhuppmann** [comment]
@sandrinecharousset @arght if I understand you correctly, there is *no agreement* between you whether a value set for time `2020-01-01T12:00+00:00` is the average of the 12 hours preceding it (morning) or the 12 hours following it (afternoon/evening). So how will a user of the openENTRANCE data know for every model how it was intended?

File: nomenclature/definitions/subannual/README.md [hunk 1]
```diff
@@ -43,6 +43,53 @@ with open(f'../subannual/months.yaml', 'r') as stream:
 mapping = dict([(m, eval(attr['duration'])) for (m, attr) in months.items()])
 ```
 
+### Days and hour data
```
Comment (omnipotent12) [inline nomenclature/definitions/subannual/README.md:46]: Should be "Daily and hourly data"

File: nomenclature/definitions/subannual/README.md [hunk 2]
```diff
@@ -43,6 +43,53 @@ with open(f'../subannual/months.yaml', 'r') as stream:
...
+
+The ISO 8601 datetime format is adopted to consider the following levels: year, month, day, hour, minute, second, and time zone.
+
+#### Example of this format:
```
Comment (omnipotent12) [inline nomenclature/definitions/subannual/README.md:50]: should be "Examples"

**Comment by sandrinecharousset** [comment]
> Many thanks for the comments. I didn't say it but my purpose was about getting attention from modelling teams to this issue and get agreements to choose a datetime format.
> I'll be checking the comments and summarize the agreements.
> The checks means that we talked about it.
> Until now, we have the following points:
> 
> * [x]  Standard datetime format     : UTC
> * [x]  Using time zone                    : no relevant
> * [x]  Using winter/summer time   : no one use
> * [ ]  Using 12-hourly resolution   : What about use am & pm format?
> 
> I'll put winter/summer time issues if some modeling teams are interested.

We do not use 12 hours resolution and I have never seen that in any model then I do not have any opinion

**Comment by erikfilias** [comment]
> @erikfilias unfortunately, about summer-time, not all modelling teams are following this discussion on GitHub - did you check with all teams that they are not using it?

In progress, I'm trying to contact them. And, I'm trying to moving the conversation to #46 .

**Comment by sandrinecharousset** [comment]
> Thanks @erikfilias!
> Can we add representative hours to this? Or is it better to think of this as a seperate issue?
> For example, in CS1 we will use something like 'm1h0' - denoting a representative (average) quantitaty in January (m1) for 00:00 - 01:00 (h0)
> 
> To me it makes the most sense to add representative hours as an extension of months.yaml, and then note its usage here. If u agree I can propose something, unless you feel it is better to have representative hours integrated with the timestamped data.

I think we may need the concept of representative period for many different cases (representative winter monday, ....), but i will leave that to concerned modelling teams

**Comment by erikfilias** [comment]
> Thanks @erikfilias!
> Can we add representative hours to this? Or is it better to think of this as a seperate issue?
> For example, in CS1 we will use something like 'm1h0' - denoting a representative (average) quantitaty in January (m1) for 00:00 - 01:00 (h0)
> 
> To me it makes the most sense to add representative hours as an extension of months.yaml, and then note its usage here. If u agree I can propose something, unless you feel it is better to have representative hours integrated with the timestamped data.

Thanks, @omnipotent12. I'm moving the conversation to  #46 .
Yeah, representative days could be added in a yaml file.
My idea is related to defined a standard datetime format that python can understand it and it could be validated without using yaml files.

**Comment by sandrinecharousset** [comment]
> @sandrinecharousset @arght if I understand you correctly, there is _no agreement_ between you whether a value set for time `2020-01-01T12:00+00:00` is the average of the 12 hours preceding it (morning) or the 12 hours following it (afternoon/evening). So how will a user of the openENTRANCE data know for every model how it was intended?

You are true....
The conventions used in tepes and plan4eu are closely related but different.... But then there is 'how it is represented in the model' and 'how it is represented in the data' which may be different. What we should work on here is agree on a convention for the data that will be on the platform. I would think that if we incorporate time series then it should either be:
* time series with a given fixed time granularity (one value per hour, or 4 values per day, 1 value per day, 1 value per week, 1 value per month...... depending on the data that we have);
* or time series over representative timesets (see discussion by @omnipotent12 )
In the first case we should not have ambiguities about which time horizon is the value related to if eg we have weeks identifiers (like Week1, Week2....), months identifiers (january....), days identifiers (12/10/2034).... which means each time a different granularity is used we need to define new identifiers (eg if we have a variable with a 2 weeks frequency, we should define identifiers for the 2-weeks periods like Fortnight1=Week1+Week2, Week1=Jan1 until jan7 included, ....); 
What do you think @erikfilias @danielhuppmann @arght @omnipotent12

**Comment by erikfilias** [comment]
> > @sandrinecharousset @arght if I understand you correctly, there is _no agreement_ between you whether a value set for time `2020-01-01T12:00+00:00` is the average of the 12 hours preceding it (morning) or the 12 hours following it (afternoon/evening). So how will a user of the openENTRANCE data know for every model how it was intended?
> 
> You are true....
> The conventions used in tepes and plan4eu are closely related but different.... But then there is 'how it is represented in the model' and 'how it is represented in the data' which may be different. What we should work on here is agree on a convention for the data that will be on the platform. I would think that if we incorporate time series then it should either be:
> 
> * time series with a given fixed time granularity (one value per hour, or 4 values per day, 1 value per day, 1 value per week, 1 value per month...... depending on the data that we have);
> * or time series over representative timesets (see discussion by @omnipotent12 )
>   In the first case we should not have ambiguities about which time horizon is the value related to if eg we have weeks identifiers (like Week1, Week2....), months identifiers (january....), days identifiers (12/10/2034).... which means each time a different granularity is used we need to define new identifiers (eg if we have a variable with a 2 weeks frequency, we should define identifiers for the 2-weeks periods like Fortnight1=Week1+Week2, Week1=Jan1 until jan7 included, ....);
>   What do you think @erikfilias @danielhuppmann @arght @omnipotent12

Please, to comment it in #46 .

**Comment by sandrinecharousset** [comment]
> Thanks @erikfilias for starting the discussion how to include continuous-time data in the common format and nomenclature.
> 
> A few issues that need clarification, in my opinion:
> 
> * do we really want that each modelling team can use any time zone? Wouldn't it be easier for data exchange to use UTC throughout, to make data more easily comparable?
> * how are modelling teams supposed to work with summer-time?
> 
> Also, how about models that do not have hourly resolution but e.g. 12-hourly resolution:
> 
> * how will a user know what the model/scenario time resolution is?
> * is the value for `2020-01-01T12:00+00:00` related to
>   
>   * the value in that hour  => I think the value for 2020-01-01T12:00+00:00 HAS TO BE the value of that hour
>   * the average over the time span 12pm until 12am => In that case we should not have 2020-01-01T12:00+00:00 in the subannual column but another identifier that explicitely represents the time span 12pm until 12am
>   * the average over the time span 6am until 6pm => In that case we should not have 2020-01-01T12:00+00:00 in the subannual column but another identifier that explicitely represents the time span 6am until 6pm
> * again summer time: should this model report its data in summer time as `2020-01-01T13:00+00:00` during summer-time (to have consistent timesteps over the year), or should the averaging be adjusted on the time-shift days? => I have no idea sorry..... never thought of that
> 
> I know that ISO 8601 **can** represent any of these issues - what we need is more guidance for the modellers on how they **should** use it in the openENTRANCE context.
> 
> @tburandt @sandrinecharousset @arght would appreciate for your point of view

**Comment by erikfilias** [comment]
S.O.S @danielhuppmann . I'm a little confused about some commits in this PR. These commits are not related to the datetime format. I'm not sure if it is a rebase problem made by me or collegues took my branch as a base and worked based on it.

**Comment by danielhuppmann** [comment]
@erikfilias I guess something went wrong during rebase - my recommended solution:

 - start a new branch based on the latest master
 - use a graphical-interface tool (Github Desktop, GitKraken) to cherry-pick the relevant commits
 - start a new PR from the new branch with a description summarizing the relevant parts of the discussion and agreement from this PR
 - close this PR with the comment "closing in favour of #<number of new PR>"

We can talk this afternoon bilaterally...

**Comment by erikfilias** [comment]
> @erikfilias I guess something went wrong during rebase - my recommended solution:
> 
> * start a new branch based on the latest master
> * use a graphical-interface tool (Github Desktop, GitKraken) to cherry-pick the relevant commits
> * start a new PR from the new branch with a description summarizing the relevant parts of the discussion and agreement from this PR
> * close this PR with the comment "closing in favour of #"
> 
> We can talk this afternoon bilaterally...

ok, great.
Thanks.

**Comment by erikfilias** [comment]
Closing in favour of #56

**Pull request closed without merging**